### PR TITLE
Audiobridge: added ability play announcement only to specific participant

### PR DESCRIPTION
**Pull request by RSATom**
or exclude some participant from announcement.

@lminiero I've tested this patch some months on older Janus version, and plan to do the same for the latest stable/master so don't want you merge it right now, but I just would like to ask Is there any chance something like this could be merged? 

Thanks in advance!

### CODE REVIEW

File: plugins/janus_audiobridge.c [hunk 1]
```diff
@@ -7756,7 +7769,8 @@ static void *janus_audiobridge_mixer_thread(void *data) {
 							buffer[i] += (resampled[i]*p->volume_gain)/100;
 						}
 					}
-				} else {
+				}
+				if(groups_num != 0) {
```
Comment (lminiero) [inline plugins/janus_audiobridge.c:7773]: Why isn't this `} else if(groups_num != 0) {`? And why doesn't this have the to check as well? Not sure why exclude isn't part of this check.

File: plugins/janus_audiobridge.c [hunk 2]
```diff
@@ -7767,9 +7781,24 @@ static void *janus_audiobridge_mixer_thread(void *data) {
...
+
+					janus_mutex_lock(&p->qmutex);
+					/* Insert packets sorting by sequence number */
+					p->inbuf = g_list_prepend(p->inbuf, pkt);
+					janus_mutex_unlock(&p->qmutex);
+				}
+
 				ps = ps->next;
 			}
-			g_list_free_full(anncs_list, (GDestroyNotify)janus_audiobridge_participant_unref);
```
Comment (lminiero) [inline plugins/janus_audiobridge.c:7772]: Why have you removed this?

File: plugins/janus_audiobridge.c [hunk 3]
```diff
@@ -7852,6 +7881,47 @@ static void *janus_audiobridge_mixer_thread(void *data) {
...
+
+#if HAVE_LIBOGG
+			for(GList *annc = anncs_list; annc; annc = annc->next) {
+				janus_audiobridge_participant *anncp = (janus_audiobridge_participant *)annc->data;
+
+				if(anncp->annc == NULL || g_atomic_int_get(&anncp->destroyed))
+					continue;
+
+				if(!anncp->annc->to && !anncp->annc->exclude)
+					continue; // already mixed
```
Comment (lminiero) [inline plugins/janus_audiobridge.c:7893]: Code style: comments should be `/*  */`

File: plugins/janus_audiobridge.c [hunk 4]
```diff
@@ -7852,6 +7881,47 @@ static void *janus_audiobridge_mixer_thread(void *data) {
...
+
+				GList *first = g_list_first(anncp->inbuf);
+				janus_audiobridge_rtp_relay_packet *pkt = (janus_audiobridge_rtp_relay_packet *)(first ? first->data : NULL);
+				curBuffer = (opus_int16 *)((pkt && pkt->length && !pkt->silence) ? pkt->data : NULL);
+				if(!curBuffer)
+					continue;
+
+				const char* to = anncp->annc->to;
+				const char* exclude = anncp->annc->exclude;
```
Comment (lminiero) [inline plugins/janus_audiobridge.c:7902]: See above on code style for pointers.

File: plugins/janus_audiobridge.c [hunk 5]
```diff
@@ -7852,6 +7881,47 @@ static void *janus_audiobridge_mixer_thread(void *data) {
...
+				if(to) {
+					if(0 == g_strcmp0(p->display, to)) {
```
Comment (lminiero) [inline plugins/janus_audiobridge.c:7904]: So `to` and `exclude` are display names? I think this is very fragile. Display names are _not_ unique, and can be changed dynamically by people in the room. If you want this to be reliable and consistent, it should be user IDs instead (which means making the same checks we do on whether they're numbers or string IDs).

**Review by lminiero — COMMENTED** [review]
Thanks for your contribution! To be honest, I'm not 100% sold on the idea and how it's implemented. Using display names is fragile, and it seems you're adding additional cycles to an already burdened loop: I feel like this could possibly be done in a lighter way, even though I haven't spent much time thinking about how. I don't fully understand why a single to and a single exclude, if not possibly to do the traditional "You're now in the room" and "Pluto has entered the room"?

File: plugins/janus_audiobridge.c [hunk 6]
```diff
@@ -1308,6 +1310,8 @@ typedef struct janus_audiobridge_file {
 	char *oggbuf;
 	gboolean started, loop;
 	gint state, headers;
+	char* to;
+	char* exclude;
```
[the remark this answers is not in the document]
Comment (RSATom) [reply plugins/janus_audiobridge.c:1314]: Sorry, just used to a little bit different code style...

File: plugins/janus_audiobridge.c (hunk 1 above)
Comment (RSATom) [reply plugins/janus_audiobridge.c:7773]: > Why isn't this } else if(groups_num != 0) {?
Just was thinking it's a little bit more clear... will fix.

> And why doesn't this have the to check as well? 
`to` handling a little bit later by add samples directly to buffers for specific user.
similar with `exclude` - it handling a little bit later by subtracting samples from buffers for specific user

P.S. I think there is another issue is there - I should take into account group notification - will fix...

File: plugins/janus_audiobridge.c (hunk 2 above)
Comment (RSATom) [reply plugins/janus_audiobridge.c:7772]: `anncs_list` is required later and moved below

File: plugins/janus_audiobridge.c (hunk 5 above)
Comment (RSATom) [reply plugins/janus_audiobridge.c:7904]: Good point. I need to think if it suitable in our case. Thanks!

**Comment by RSATom** [comment]
> I don't fully understand why a single to and a single exclude, if not possibly to do the traditional "You're now in the room" and "Pluto has entered the room"?

`to` was added just as counterpart to `exclude`, just to have symmetry in API.
What about `exclude` - it's for a little bit more complex task. It's using it for realtime text-to-speach (with Google TTS API) for messages in chat (and for text-to-speach translated to different languages) - and user could want not hear it's own messages.

**Pull request closed without merging**